File: bench/coquic-go-perf/internal/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"time"

	"github.com/minhuw/coquic/bench/coquic-go-perf/internal/config"
)

const (
	ProtocolVersionLegacy       uint32 = 1
	ProtocolVersionMilliseconds uint32 = 2
	ProtocolVersion             uint32 = 3
	ControlStreamID             uint64 = 0
	FirstDataStreamID           uint64 = 4
)

const (
	messageSessionStart    byte = 1
	messageSessionReady    byte = 2
	messageSessionError    byte = 3
	messageSessionComplete byte = 4
)

const (
	optionalTotalBytesFlag byte = 0x01
	optionalRequestsFlag   byte = 0x02
)

type SessionStart struct {
	ProtocolVersion  uint32
	Mode             config.Mode
	Direction        config.Direction
	RequestBytes     uint64
	ResponseBytes    uint64
	TotalBytes       *uint64
	Requests         *uint64
	Warmup           time.Duration
	Duration         time.Duration
	Streams          uint64
	Connections      uint64
	RequestsInFlight uint64
}

type SessionReady struct {
	ProtocolVersion uint32
}

type SessionError struct {
	Reason string
}

type SessionComplete struct {
	BytesSent         uint64
	BytesReceived     uint64
	RequestsCompleted uint64
}

type ControlMessage interface{}

func EncodeControlMessage(message ControlMessage) []byte {
	payload := make([]byte, 0, 96)
	var messageType byte
	switch value := message.(type) {
	case SessionStart:
		appendU32(&payload, value.ProtocolVersion)
		appendU8(&payload, modeToU8(value.Mode))
		appendU8(&payload, directionToU8(value.Direction))
		appendU64(&payload, value.RequestBytes)
		appendU64(&payload, value.ResponseBytes)
		if value.ProtocolVersion != ProtocolVersionLegacy {
			appendU8(&payload, sessionStartOptionalFlags(value))
		}
		appendU64(&payload, optionalValue(value.TotalBytes))
		appendU64(&payload, optionalValue(value.Requests))
		if value.ProtocolVersion == ProtocolVersionLegacy ||
			value.ProtocolVersion == ProtocolVersionMilliseconds {
			appendU64(&payload, durationMillis(value.Warmup))
			appendU64(&payload, durationMillis(value.Duration))
		} else {
			appendU64(&payload, durationMicros(value.Warmup))
			appendU64(&payload, durationMicros(value.Duration))
		}
		appendU64(&payload, value.Streams)
		appendU64(&payload, value.Connections)
		appendU64(&payload, value.RequestsInFlight)
		messageType = messageSessionStart
	case SessionReady:
		appendU32(&payload, value.ProtocolVersion)
		messageType = messageSessionReady
	case SessionError:
		appendU32(&payload, uint32(len(value.Reason)))
		payload = append(payload, []byte(value.Reason)...)
		messageType = messageSessionError
	case SessionComplete:
		appendU64(&payload, value.BytesSent)
		appendU64(&payload, value.BytesReceived)
		appendU64(&payload, value.RequestsCompleted)
		messageType = messageSessionComplete
	default:
		return nil
	}

	out := make([]byte, 0, 5+len(payload))
	appendU8(&out, messageType)
	appendU32(&out, uint32(len(payload)))
	out = append(out, payload...)
	return out
}

func DecodeControlMessage(input []byte) (ControlMessage, bool) {
	messageType, input, ok := takeU8(input)
	if !ok {
		return nil, false
	}
	payloadSize, input, ok := takeU32(input)
	if !ok || len(input) != int(payloadSize) {
		return nil, false
	}

	switch messageType {
	case messageSessionStart:
		return decodeSessionStart(input)
	case messageSessionReady:
		protocolVersion, rest, ok := takeU32(input)
		if !ok || len(rest) != 0 {
			return nil, false
		}
		return SessionReady{ProtocolVersion: protocolVersion}, true
	case messageSessionError:
		reason, rest, ok := takeString(input)
		if !ok || len(rest) != 0 {
			return nil, false
		}
		return SessionError{Reason: reason}, true
	case messageSessionComplete:
		bytesSent, rest, ok := takeU64(input)
		if !ok {
			return nil, false
		}
		bytesReceived, rest, ok := takeU64(rest)
		if !ok {
			return nil, false
		}
		requestsCompleted, rest, ok := takeU64(rest)
		if !ok || len(rest) != 0 {
			return nil, false
		}
		return SessionComplete{
			BytesSent:         bytesSent,
			BytesReceived:     bytesReceived,
			RequestsCompleted: requestsCompleted,
		}, true
	default:
		return nil, false
	}
}

func TakeControlMessage(buffer *[]byte) (ControlMessage, bool) {
	if len(*buffer) < 5 {
		return nil, false
	}
	payloadSize := binary.BigEndian.Uint32((*buffer)[1:5])
	frameSize := 5 + int(payloadSize)
	if len(*buffer) < frameSize {
		return nil, false
	}
	frame := append([]byte(nil), (*buffer)[:frameSize]...)
	*buffer = (*buffer)[frameSize:]
	return DecodeControlMessage(frame)
}

func NextClientStreamID(current uint64) uint64 {
	if current == 0 {
		return FirstDataStreamID
	}
	return current + 4
}

func decodeSessionStart(input []byte) (ControlMessage, bool) {
	protocolVersion, input, ok := takeU32(input)
	if !ok {
		return nil, false
	}
	modeRaw, input, ok := takeU8(input)
	if !ok {
		return nil, false
	}
	mode, ok := parseMode(modeRaw)
	if !ok {
		return nil, false
	}
	directionRaw, input, ok := takeU8(input)
	if !ok {
		return nil, false
	}
	direction, ok := parseDirection(directionRaw)
	if !ok {
		return nil, false
	}
	requestBytes, input, ok := takeU64(input)
	if !ok {
		return nil, false
	}
	responseBytes, input, ok := takeU64(input)
	if !ok {
		return nil, false
	}

	var optionalFlags byte
	if protocolVersion == ProtocolVersion || protocolVersion == ProtocolVersionMilliseconds {
		optionalFlags, input, ok = takeU8(input)
		if !ok {
			return nil, false
		}
	} else if protocolVersion != ProtocolVersionLegacy {
		return nil, false
	}

	totalBytesRaw, input, ok := takeU64(input)
	if !ok {
		return nil, false
	}
	requestsRaw, input, ok := takeU64(input)
	if !ok {
		return nil, false
	}
	warmupRaw, input, ok := takeU64(input)
	if !ok {
		return nil, false
	}
	durationRaw, input, ok := takeU64(input)
	if !ok {
		return nil, false
	}
	streams, input, ok := takeU64(input)
	if !ok {
		return nil, false
	}
	connections, input, ok := takeU64(input)
	if !ok {
		return nil, false
	}
	requestsInFlight, input, ok := takeU64(input)
	if !ok || len(input) != 0 {
		return nil, false
	}

	var totalBytes *uint64
	var requests *uint64
	if protocolVersion == ProtocolVersionLegacy {
		if totalBytesRaw != 0 {
			totalBytes = ptr(totalBytesRaw)
		}
		if requestsRaw != 0 {
			requests = ptr(requestsRaw)
		}
	} else {
		if optionalFlags&optionalTotalBytesFlag != 0 {
			totalBytes = ptr(totalBytesRaw)
		}
		if optionalFlags&optionalRequestsFlag != 0 {
			requests = ptr(requestsRaw)
		}
	}

	warmup := time.Duration(warmupRaw) * time.Microsecond
	duration := time.Duration(durationRaw) * time.Microsecond
	if protocolVersion == ProtocolVersionLegacy ||
		protocolVersion == ProtocolVersionMilliseconds {
		warmup = time.Duration(warmupRaw) * time.Millisecond
		duration = time.Duration(durationRaw) * time.Millisecond
	}

	return SessionStart{
		ProtocolVersion:  protocolVersion,
		Mode:             mode,
		Direction:        direction,
		RequestBytes:     requestBytes,
		ResponseBytes:    responseBytes,
		TotalBytes:       totalBytes,
		Requests:         requests,
		Warmup:           warmup,
		Duration:         duration,
		Streams:          streams,
		Connections:      connections,
		RequestsInFlight: requestsInFlight,
	}, true
}

func appendU8(out *[]byte, value byte) {
	*out = append(*out, value)
}

func appendU32(out *[]byte, value uint32) {
	var bytes [4]byte
	binary.BigEndian.PutUint32(bytes[:], value)
	*out = append(*out, bytes[:]...)
}

func appendU64(out *[]byte, value uint64) {
	var bytes [8]byte
	binary.BigEndian.PutUint64(bytes[:], value)
	*out = append(*out, bytes[:]...)
}

func takeU8(input []byte) (byte, []byte, bool) {
	if len(input) < 1 {
		return 0, input, false
	}
	return input[0], input[1:], true
}

func takeU32(input []byte) (uint32, []byte, bool) {
	if len(input) < 4 {
		return 0, input, false
	}
	return binary.BigEndian.Uint32(input[:4]), input[4:], true
}

func takeU64(input []byte) (uint64, []byte, bool) {
	if len(input) < 8 {
		return 0, input, false
	}
	return binary.BigEndian.Uint64(input[:8]), input[8:], true
}

func takeString(input []byte) (string, []byte, bool) {
	size, input, ok := takeU32(input)
	if !ok || len(input) < int(size) {
		return "", input, false
	}
	return string(input[:size]), input[size:], true
}

func modeToU8(mode config.Mode) byte {
	switch mode {
	case config.ModeBulk:
		return 0
	case config.ModeRR:
		return 1
	case config.ModeCRR:
		return 2
	case config.ModePersistentRR:
		return 3
	default:
		return 0
	}
}

func parseMode(value byte) (config.Mode, bool) {
	switch value {
	case 0:
		return config.ModeBulk, true
	case 1:
		return config.ModeRR, true
	case 2:
		return config.ModeCRR, true
	case 3:
		return config.ModePersistentRR, true
	default:
		return 0, false
	}
}

func directionToU8(direction config.Direction) byte {
	switch direction {
	case config.DirectionUpload:
		return 0
	case config.DirectionDownload:
		return 1
	default:
		return 0
	}
}

func parseDirection(value byte) (config.Direction, bool) {
	switch value {
	case 0:
		return config.DirectionUpload, true
	case 1:
		return config.DirectionDownload, true
	default:
		return 0, false
	}
}

func sessionStartOptionalFlags(start SessionStart) byte {
	var flags byte
	if start.TotalBytes != nil {
		flags |= optionalTotalBytesFlag
	}
	if start.Requests != nil {
		flags |= optionalRequestsFlag
	}
	return flags
}

func optionalValue(value *uint64) uint64 {
	if value == nil {
		return 0
	}
	return *value
}

func durationMicros(duration time.Duration) uint64 {
	return uint64(duration / time.Microsecond)
}

func durationMillis(duration time.Duration) uint64 {
	return uint64(duration / time.Millisecond)
}

func ptr(value uint64) *uint64 {
	return &value
}

File: bench/coquic-rust-perf/src/protocol.rs
use crate::config::{Direction, Mode};
use std::time::Duration;

pub const PROTOCOL_VERSION_LEGACY: u32 = 1;
pub const PROTOCOL_VERSION_MILLISECONDS: u32 = 2;
pub const PROTOCOL_VERSION: u32 = 3;
pub const CONTROL_STREAM_ID: u64 = 0;
pub const FIRST_DATA_STREAM_ID: u64 = 4;

const MESSAGE_SESSION_START: u8 = 1;
const MESSAGE_SESSION_READY: u8 = 2;
const MESSAGE_SESSION_ERROR: u8 = 3;
const MESSAGE_SESSION_COMPLETE: u8 = 4;

const OPTIONAL_TOTAL_BYTES_FLAG: u8 = 0x01;
const OPTIONAL_REQUESTS_FLAG: u8 = 0x02;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SessionStart {
    pub protocol_version: u32,
    pub mode: Mode,
    pub direction: Direction,
    pub request_bytes: u64,
    pub response_bytes: u64,
    pub total_bytes: Option<u64>,
    pub requests: Option<u64>,
    pub warmup: Duration,
    pub duration: Duration,
    pub streams: u64,
    pub connections: u64,
    pub requests_in_flight: u64,
}

impl Default for SessionStart {
    fn default() -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            mode: Mode::Bulk,
            direction: Direction::Download,
            request_bytes: 0,
            response_bytes: 0,
            total_bytes: None,
            requests: None,
            warmup: Duration::ZERO,
            duration: Duration::ZERO,
            streams: 1,
            connections: 1,
            requests_in_flight: 1,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SessionReady {
    pub protocol_version: u32,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SessionError {
    pub reason: String,
}

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct SessionComplete {
    pub bytes_sent: u64,
    pub bytes_received: u64,
    pub requests_completed: u64,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ControlMessage {
    SessionStart(SessionStart),
    SessionReady(SessionReady),
    SessionError(SessionError),
    SessionComplete(SessionComplete),
}

pub fn encode_control_message(message: &ControlMessage) -> Vec<u8> {
    let mut payload = Vec::new();
    let message_type = match message {
        ControlMessage::SessionStart(start) => {
            append_u32(&mut payload, start.protocol_version);
            append_u8(&mut payload, mode_to_u8(start.mode));
            append_u8(&mut payload, direction_to_u8(start.direction));
            append_u64(&mut payload, start.request_bytes);
            append_u64(&mut payload, start.response_bytes);
            if start.protocol_version != PROTOCOL_VERSION_LEGACY {
                append_u8(&mut payload, session_start_optional_flags(start));
            }
            append_u64(&mut payload, start.total_bytes.unwrap_or(0));
            append_u64(&mut payload, start.requests.unwrap_or(0));
            if start.protocol_version == PROTOCOL_VERSION_LEGACY
                || start.protocol_version == PROTOCOL_VERSION_MILLISECONDS
            {
                append_u64(&mut payload, duration_millis(start.warmup));
                append_u64(&mut payload, duration_millis(start.duration));
            } else {
                append_u64(&mut payload, duration_micros(start.warmup));
                append_u64(&mut payload, duration_micros(start.duration));
            }
            append_u64(&mut payload, start.streams);
            append_u64(&mut payload, start.connections);
            append_u64(&mut payload, start.requests_in_flight);
            MESSAGE_SESSION_START
        }
        ControlMessage::SessionReady(ready) => {
            append_u32(&mut payload, ready.protocol_version);
            MESSAGE_SESSION_READY
        }
        ControlMessage::SessionError(error) => {
            append_u32(&mut payload, error.reason.len() as u32);
            payload.extend_from_slice(error.reason.as_bytes());
            MESSAGE_SESSION_ERROR
        }
        ControlMessage::SessionComplete(complete) => {
            append_u64(&mut payload, complete.bytes_sent);
            append_u64(&mut payload, complete.bytes_received);
            append_u64(&mut payload, complete.requests_completed);
            MESSAGE_SESSION_COMPLETE
        }
    };

    let mut out = Vec::with_capacity(5 + payload.len());
    append_u8(&mut out, message_type);
    append_u32(&mut out, payload.len() as u32);
    out.extend_from_slice(&payload);
    out
}

pub fn decode_control_message(mut input: &[u8]) -> Option<ControlMessage> {
    let message_type = take_u8(&mut input)?;
    let payload_size = take_u32(&mut input)? as usize;
    if input.len() != payload_size {
        return None;
    }

    match message_type {
        MESSAGE_SESSION_START => decode_session_start(input).map(ControlMessage::SessionStart),
        MESSAGE_SESSION_READY => {
            let protocol_version = take_u32(&mut input)?;
            input
                .is_empty()
                .then_some(ControlMessage::SessionReady(SessionReady {
                    protocol_version,
                }))
        }
        MESSAGE_SESSION_ERROR => {
            let reason = take_string(&mut input)?;
            input
                .is_empty()
                .then_some(ControlMessage::SessionError(SessionError { reason }))
        }
        MESSAGE_SESSION_COMPLETE => {
            let bytes_sent = take_u64(&mut input)?;
            let bytes_received = take_u64(&mut input)?;
            let requests_completed = take_u64(&mut input)?;
            input
                .is_empty()
                .then_some(ControlMessage::SessionComplete(SessionComplete {
                    bytes_sent,
                    bytes_received,
                    requests_completed,
                }))
        }
        _ => None,
    }
}

pub fn take_control_message(buffer: &mut Vec<u8>) -> Option<ControlMessage> {
    if buffer.len() < 5 {
        return None;
    }

    let payload_size = u32::from_be_bytes([buffer[1], buffer[2], buffer[3], buffer[4]]) as usize;
    let frame_size = 5 + payload_size;
    if buffer.len() < frame_size {
        return None;
    }

    let frame: Vec<u8> = buffer.drain(..frame_size).collect();
    decode_control_message(&frame)
}

pub fn next_client_stream_id(current: u64) -> u64 {
    if current == 0 {
        FIRST_DATA_STREAM_ID
    } else {
        current + 4
    }
}

fn decode_session_start(mut input: &[u8]) -> Option<SessionStart> {
    let protocol_version = take_u32(&mut input)?;
    let mode = parse_mode(take_u8(&mut input)?)?;
    let direction = parse_direction(take_u8(&mut input)?)?;
    let request_bytes = take_u64(&mut input)?;
    let response_bytes = take_u64(&mut input)?;

    let optional_flags = if protocol_version == PROTOCOL_VERSION
        || protocol_version == PROTOCOL_VERSION_MILLISECONDS
    {
        take_u8(&mut input)?
    } else if protocol_version == PROTOCOL_VERSION_LEGACY {
        0
    } else {
        return None;
    };

    let total_bytes_raw = take_u64(&mut input)?;
    let requests_raw = take_u64(&mut input)?;
    let warmup_raw = take_u64(&mut input)?;
    let duration_raw = take_u64(&mut input)?;
    let streams = take_u64(&mut input)?;
    let connections = take_u64(&mut input)?;
    let requests_in_flight = take_u64(&mut input)?;
    if !input.is_empty() {
        return None;
    }

    let (total_bytes, requests) = if protocol_version == PROTOCOL_VERSION_LEGACY {
        (
            (total_bytes_raw != 0).then_some(total_bytes_raw),
            (requests_raw != 0).then_some(requests_raw),
        )
    } else {
        (
            ((optional_flags & OPTIONAL_TOTAL_BYTES_FLAG) != 0).then_some(total_bytes_raw),
            ((optional_flags & OPTIONAL_REQUESTS_FLAG) != 0).then_some(requests_raw),
        )
    };

    let (warmup, duration) = if protocol_version == PROTOCOL_VERSION_LEGACY
        || protocol_version == PROTOCOL_VERSION_MILLISECONDS
    {
        (
            Duration::from_millis(warmup_raw),
            Duration::from_millis(duration_raw),
        )
    } else {
        (
            Duration::from_micros(warmup_raw),
            Duration::from_micros(duration_raw),
        )
    };

    Some(SessionStart {
        protocol_version,
        mode,
        direction,
        request_bytes,
        response_bytes,
        total_bytes,
        requests,
        warmup,
        duration,
        streams,
        connections,
        requests_in_flight,
    })
}

fn append_u8(out: &mut Vec<u8>, value: u8) {
    out.push(value);
}

fn append_u32(out: &mut Vec<u8>, value: u32) {
    out.extend_from_slice(&value.to_be_bytes());
}

fn append_u64(out: &mut Vec<u8>, value: u64) {
    out.extend_from_slice(&value.to_be_bytes());
}

fn take_u8(input: &mut &[u8]) -> Option<u8> {
    let (head, tail) = input.split_first()?;
    *input = tail;
    Some(*head)
}

fn take_u32(input: &mut &[u8]) -> Option<u32> {
    if input.len() < 4 {
        return None;
    }
    let value = u32::from_be_bytes(input[..4].try_into().ok()?);
    *input = &input[4..];
    Some(value)
}

fn take_u64(input: &mut &[u8]) -> Option<u64> {
    if input.len() < 8 {
        return None;
    }
    let value = u64::from_be_bytes(input[..8].try_into().ok()?);
    *input = &input[8..];
    Some(value)
}

fn take_string(input: &mut &[u8]) -> Option<String> {
    let size = take_u32(input)? as usize;
    if input.len() < size {
        return None;
    }
    let value = String::from_utf8(input[..size].to_vec()).ok()?;
    *input = &input[size..];
    Some(value)
}

fn mode_to_u8(mode: Mode) -> u8 {
    match mode {
        Mode::Bulk => 0,
        Mode::Rr => 1,
        Mode::Crr => 2,
        Mode::PersistentRr => 3,
    }
}

fn parse_mode(value: u8) -> Option<Mode> {
    match value {
        0 => Some(Mode::Bulk),
        1 => Some(Mode::Rr),
        2 => Some(Mode::Crr),
        3 => Some(Mode::PersistentRr),
        _ => None,
    }
}

fn direction_to_u8(direction: Direction) -> u8 {
    match direction {
        Direction::Upload => 0,
        Direction::Download => 1,
    }
}

fn parse_direction(value: u8) -> Option<Direction> {
    match value {
        0 => Some(Direction::Upload),
        1 => Some(Direction::Download),
        _ => None,
    }
}

fn session_start_optional_flags(start: &SessionStart) -> u8 {
    let mut flags = 0;
    if start.total_bytes.is_some() {
        flags |= OPTIONAL_TOTAL_BYTES_FLAG;
    }
    if start.requests.is_some() {
        flags |= OPTIONAL_REQUESTS_FLAG;
    }
    flags
}

fn duration_micros(duration: Duration) -> u64 {
    duration.as_micros().try_into().unwrap_or(u64::MAX)
}

fn duration_millis(duration: Duration) -> u64 {
    duration.as_millis().try_into().unwrap_or(u64::MAX)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trips_v3_session_start() {
        let start = SessionStart {
            protocol_version: PROTOCOL_VERSION,
            mode: Mode::Rr,
            direction: Direction::Download,
            request_bytes: 64,
            response_bytes: 128,
            total_bytes: Some(1024),
            requests: Some(10),
            warmup: Duration::from_millis(250),
            duration: Duration::from_secs(2),
            streams: 4,
            connections: 2,
            requests_in_flight: 3,
        };
        let message = ControlMessage::SessionStart(start.clone());
        let encoded = encode_control_message(&message);
        assert_eq!(decode_control_message(&encoded), Some(message));
    }

    #[test]
    fn take_control_message_waits_for_complete_frame() {
        let encoded = encode_control_message(&ControlMessage::SessionReady(SessionReady {
            protocol_version: PROTOCOL_VERSION,
        }));
        let mut partial = encoded[..encoded.len() - 1].to_vec();
        assert_eq!(take_control_message(&mut partial), None);

        let mut complete = encoded;
        assert_eq!(
            take_control_message(&mut complete),
            Some(ControlMessage::SessionReady(SessionReady {
                protocol_version: PROTOCOL_VERSION
            }))
        );
        assert!(complete.is_empty());
    }

    #[test]
    fn next_client_stream_ids_match_perf_layout() {
        assert_eq!(next_client_stream_id(0), 4);
        assert_eq!(next_client_stream_id(4), 8);
    }
}
